R/harmonicHMC.R

```diff
@@ -0,0 +1,106 @@
+#' Run Harmonic HMC Sampler
+#'
+#' Sample from a truncated Gaussian distribution with constraints Fx+g >= 0.
+#'
+#' @param n number of samples
+#' @param initialPosition starting value for parameters
+#' @param constraintDirec F matrix (k-by-d matrix where k is the number of
+#' linear constraints)
+#' @param constraintBound g vector (k dimensional)
+#' @param choleskyFactor upper triangular matrix R from cholesky decomposition
+#' of precision or covariance matrix into R^TR
+#' @param unconstrainedMean mean of unconstrained Gaussian
+#' @param precParametrized boolean for whether parametrization is by precision
+#' (TRUE) or covariance matrix (FALSE)
+#' @param totalTime amount of time the particle bounces for each sample. Can
+#' either be a scalar value for a fixed time across all samples, or a length 2
+#' vector of a lower and upper bound for uniform distribution from which the
+#' bounce time is drawn from for each sample.
+#' @param seed random seed
+#' @param diagnosticMode boolean for whether to return the bounce distances for
+#' each sample
+#' @return List of
+#' "samples": d x n matrix of samples
+#' "bounceDistances": list of bounces for each sample, only present if
+#' diagnosticMode is TRUE
+#' @export
+#'
+#' @examples
+#' set.seed(1)
+#' d = 10
+#' A = matrix(runif(d^2)*2-1, ncol=d)
+#' Sigma = t(A) %*% A
+#' R = cholesky(Sigma)
+#' mu = rep(0,d)
+#' constraintDirec = diag(d)
+#' constraintBound = rep(0,d)
+#' initial = rep(1, d)
+#' results = runHHMC(
+#' 100,
+#' initial,
+#' constraintDirec,
+#' constraintBound,
+#' R,
+#' mu,
+#' precParametrized = FALSE
+#' )
+
+runHHMC = function(n,
+                   initialPosition,
+                   constraintDirec,
+                   constraintBound,
+                   choleskyFactor,
+                   unconstrainedMean,
+                   precParametrized = TRUE,
+                   totalTime = c(pi / 8, pi / 2),
+                   seed = 1,
+                   diagnosticMode = FALSE) {
+  set.seed(seed)
+  samples = matrix(nrow = ncol(constraintDirec), ncol = n)
+  randomBounceTime = ifelse(length(totalTime)==2, TRUE, FALSE)
+  bounceDistances = vector(mode = "list",
+                           length = ifelse(diagnosticMode, n, 0))
+  whitenedConstraints = applyWhitenTransform(
+    constraintDirec,
+    constraintBound,
+    choleskyFactor,
+    unconstrainedMean,
+    precParametrized
+  )
+  position = whitenPosition(
+    initialPosition,
+    constraintDirec,
+    constraintBound,
+    choleskyFactor,
+    unconstrainedMean,
+    precParametrized
+  )
+  for (i in 1:n) {
+    bounceTime = ifelse(randomBounceTime, 
```

Comment:
This isn't the most computationally efficient since we evaluate the ifelse every iteration, so for low dimension high samples it adds some time. My thinking is that the total time is already low enough it doesn't matter (e.g. in 10 dimensions, 100k samples taking a half second longer isn't too bad when the total run time is still ~5 seconds).

Reply:
It seems `runif` handles lower_bound = upper_bound by returning the fixed value. So we can always use `runif` abd avoid ifelse?

Reply:
I think technically doing `runif(1, x, x)` is slower than `ifelse(condition, runif(), constant)` since the it still tries to simulate from a Unif[x,x] distribution, though it's pretty marginal:

```
> ptm = proc.time(); for(i in 1:100000){ifelse(TRUE, 1, 2)}; proc.time() - ptm
   user  system elapsed 
  0.129   0.000   0.129 
> ptm = proc.time(); for(i in 1:100000){runif(1,1,1)}; proc.time() - ptm
   user  system elapsed 
  0.193   0.000   0.192 
```

So happy to go either way - just doing runif without ifelse saves a few lines (but not sure it's clearer)

Reply:
Yes, if given a scalar, you can just make it a vector of repeated values and always use `runif`. 

Btw, `bounceTime` is a misnomer.

Reply:
Refactored